### DISCUSSION TITLE
PlantUML Build / Release Process

### DISCUSSION THREAD

**Discussion by matthew16550** [plantuml/plantuml#653] [General]
The PlantUML build / release process has evolved over many years.  Several automatic & manual systems are involved.

This is a place to discuss simplifying & automating some more.

(There are earlier discussions in #139, #486, #557)

**Comment by matthew16550** [comment]
As a start, here is my attempt to list all the main pieces of the puzzle.  (Will update as more are found)

## Source Repositories

* [plantuml (Sourceforge)](https://sourceforge.net/p/plantuml/code/HEAD/tree/)
  * Subversion
  * This is the main development repo
* [plantuml (GitHub)](https://github.com/plantuml/plantuml)
  * Manually synced with Sourceforge
* [plantuml-epl](https://github.com/plantuml/plantuml-epl)
* [plantuml-mit](https://github.com/plantuml/plantuml-mit)
* [plantuml-server](https://github.com/plantuml/plantuml-server)
* [plantuml-stdlib](https://github.com/plantuml/plantuml-stdlib)

# Continuous Integration

[GitHub](https://github.com/plantuml/plantuml/actions/workflows/ci.yml)
* Builds & tests every PR & every commit on master branch
* Creates a [GitHub Release](https://github.com/plantuml/plantuml/releases) for every ["v" tag](https://github.com/plantuml/plantuml/blob/master/docs/releasing.md) (this is new functionality and hasn't been used yet)

[Travis](https://travis-ci.org/github/plantuml)

* Has not run any builds since June 2021
* Assume we would need to apply for [OSS credits](https://docs.travis-ci.com/user/billing-faq/#what-if-i-am-building-open-source) to continue using it

# Artifact Repositories

[Docker Hub](https://hub.docker.com/r/plantuml)
* `plantuml-server` container

[GitHub plantuml releases](https://github.com/plantuml/plantuml/releases) 
* jar, javadoc, sources as built by Maven

[GitHub plantuml-server releases](https://github.com/plantuml/plantuml-server/releases)
* plantuml-server war

[Maven Central](https://repo1.maven.org/maven2/net/sourceforge/plantuml/)
* jar, javadoc, sources as built by Maven

[Sourceforge](https://sourceforge.net/projects/plantuml/files/)
* jar
* nodot jar
* war
* jars in `.zip` with various licenses
* sources in `.tar.gz` with various licenses

**Comment by matthew16550** [comment]
[New public beta of GitHub Releases](https://github.blog/2021-10-04-beta-github-releases-improving-release-experience/)

Currently, users will need to enable it in [Feature Preview](https://docs.github.com/en/get-started/using-github/exploring-early-access-releases-with-feature-preview).

I think the big thing for us is that tags no longer show in the Release list view.

Also, Release Notes can now be auto-generated.  Not clear if that will help us, ~~or if it can be part of the CI workflow~~.

EDIT: There is an API for auto-generated release notes, the docs are still being written but a [draft](https://github.com/github/feedback/discussions/5962#discussioncomment-1426713) is available.  So I assume it could be part of our CI.

**Comment by The-Lum** [comment]
Hello @matthew16550, and all,

Linked to:
- #557

Is it possible to have directly `plantuml.jar` (on artefact list) or another artefact with only the `plantuml.jar`, even for [intermediate CI](https://github.com/plantuml/plantuml/actions) and not only for [Release](https://github.com/plantuml/plantuml/releases)?
_Especially to test intermediate version, without download the full **artefact** and extract only the minimal `plantuml.jar`._

And same question as [mine](https://github.com/plantuml/plantuml/issues/557#issuecomment-955209058) on #557: 
- How to retrieve the URL of the latest **artefact**?
_[Hints on [stackoverflow](https://stackoverflow.com/questions/60789862/url-of-the-last-artifact-of-a-github-action-build) or [comments on github.community](https://github.community/t/get-artifact-url-in-steps-after-upload-artifact/17463)]_

Regards.

**Reply by matthew16550** [reply]
From what I have read it looks like the zip is auto-created during download.  The CI workflow already uploads separate jar files but GitHub does not provide separate downloads.  There is some discussion here: https://github.com/actions/upload-artifact/issues/39

We might be able to use [GitHub Packages](https://docs.github.com/en/packages) to make the jar available, I can try in the next few weeks.

**Reply by The-Lum** [reply]
Hello @matthew16550 and all,

Would it be possible to add,  on the file `ci.yml`, just after the current **'Upload jar artifacts'**:
```yml
      - name: Upload jar artifact (only simple jar: without javadoc or sources)
        if: ${{ matrix.release_from_this_build }}
        uses: actions/upload-artifact@v2
        with:
          name: plantuml-actions-artifact-jar
          path: target/*SNAPSHOT.jar
```

Can I make a PR? or have you any ideas to retrieve the value of the `pom-version` _(just to avoid to hard-code `*SNAPSHOT.jar`, just to retrieve the plantuml jar (without javadoc or sources))_?

Regards.

**Reply by matthew16550** [reply]
Try it in a forked repo 😄.  I suspect when we download, it will be a zip containing one jar so we will still need to unzip it.

Something like this in an earlier shell step should get the POM version.

    mvn -q -Dexec.executable=echo -Dexec.args='${project.version}' --non-recursive exec:exec

A few days ago I was experimenting publishing snapshots in docker containers to GitHub Packages, it seems to work ok but is quite complicated.

Looked at containers first because GitHub Packages lets us make them publicly available so your repo can use them without any permission complications.  But I think all the docker complication could make it a bad tradeoff.  

An alternative I will look at soon is publishing snapshot jars to GitHub Packages and making a personal access token so your repo has access to them.

Perhaps if we wait a few months Github will add a permissions feature to make sharing jars easier.

**Reply by matthew16550** [reply]
I just thought of a simple solution, we could make a `SNAPSHOT` release like here: https://github.com/matthew16550/plantuml/releases

It is marked as `pre-release` so on the [main repo page](https://github.com/matthew16550/plantuml) the Releases section at the right still shows the latest real release, the snapshot does not appear there.

I made those releases in my repo manually to show what it would look like but I'm fairly confident we can do it in CI.

**Disadvantages**
* The jar is available for anyone to download but we do not want most people using it.
* Every time we make a new snapshot there will be a brief period while the release is being updated so is not available  (I guess only a few seconds).

@arnaudroques what do you think?  (Given more time I can probably find a way for GitHub Packages to make jars only available to approved persons - if we are concerned about that)

**Reply by arnaudroques** [reply]
>     The jar is available for anyone to download but we do not want most people using it.

That's not a real issue to me. As long as there are some warning (**This is a pre-release of the latest development work. It is not ready for general use**.) that sounds great to me.

As this looks like the easiest solution, I vote for it!

**Reply by matthew16550** [reply]
@The-Lum will a SNAPSHOT release work for you?

**Reply by The-Lum** [reply]
But make a `pre-release` each time the CI runs it not good... _(it will pollute the releases)_

**Reply by arnaudroques** [reply]
@The-Lum You're right in theory. However, if you have a look here https://github.com/matthew16550/plantuml/releases I found this is quite clear.
CI will run only when commit in the master, right ?

**Reply by matthew16550** [reply]
Yes only on the `master` branch.  To avoid any pollution I will try editing the existing release to attach the new jar or deleting the old release first.

**Reply by The-Lum** [reply]
As a reminder, the need is to have each time the CI runs (on the `master` branch of `plantuml/plantuml`) a single `plantuml[.*].jar` file, only for testing, awaiting each times the next release.

Today we have an artifact (`<run_number>-jars.zip` on [actions > runs > artifacts](https://github.com/plantuml/plantuml/actions/)) that contains too much stuff (jar + src + javadoc), and I would like to avoid downloading unnecessary and bulky things.
But I can be satisfied with this one and extract on the `plantuml-<version>-SNAPSHOT.jar` from the big archive. But I would have downloaded things for nothing...

**Reply by matthew16550** [reply]
@arnaudroques over for you to decide.  I'm happy to try the snapshot release, think it will not take much time.  It's kind of a nice thing to have as other people may find it useful.  But not high priority.

**Reply by arnaudroques** [reply]
@matthew16550 Let's go for it. I think it will be a nice feature for advanced users since they won't have to wait for official releases to try new stuff. (Even though the [official webserver](http://www.plantuml.com/plantuml/uml/SyfFKj2rKt3CoKnELR1Io4ZDoSa70000) often implements new features before official releases).

Keep all this simple (e.g. _delete the old release first_). We do not need a 99.9999% SLA here :-)

If people get confused (which I doubt) we would remove this snapshot/pre-release section.

**Reply by The-Lum** [reply]
In parallel...
Awaiting `pre-release`, I create this PR:
- #764

And that also automatic...  
See direct result 🎉 on:
- https://github.com/plantuml/plantuml/actions/runs/1453566462
_[even if that is a `jar` on a `zip`, that is a good step until `pre-release`...]_

**Reply by matthew16550** [reply]
I have it working in CI, https://github.com/matthew16550/plantuml/releases

Need to test a little more to make sure I haven't broken the main release step then will make a PR.

**Reply by matthew16550** [reply]
@The-Lum the `snapshot` releasing is merged.

Hopefully you can download from https://github.com/plantuml/plantuml/releases/download/snapshot/plantuml-SNAPSHOT.jar

**Reply by The-Lum** [reply]
Thank you very much, 
just have to test in the coming days...

We will now be able to have all the means to work, test and make new themes....

**Reply by matthew16550** [reply]
Hello @The-Lum
Are you using the SNAPSHOT releases now?
Is it ok if I remove the `Upload jar artifact (only simple jar - without javadoc or sources)` step as part of #769  ?

**Reply by The-Lum** [reply]
Hello @matthew16550 

Yes you can... I use now the snapshot.
Your snapshot method is more better than my method with artefact (jar on a zip!).

Thanks a lot.

**Reply by The-Lum** [reply]
It is possible to add a timestamp file on the snapshot.
_I can make a PR..._

Regards.

**Reply by The-Lum** [reply]
FYI @matthew16550, I just post a PR... see:
- #799

**Reply by matthew16550** [reply]
Hello @The-Lum,

I would like to rename the timestamp file to `plantuml-SNAPSHOT.timestamp.lock` or maybe just `plantuml-SNAPSHOT.timestamp`?

The reason is for #769 I am adding `.pom` & `.asc` signature files to the snapshot release so there will be quite a lot of files and if we change the filename to `... dot timestamp` then the list of files in the release will alpha sort more nicely, see this [example](https://github.com/matthew16550/plantuml/releases/tag/snapshot) in my forked repo.

We add the files with the `gh release create` command, I tried listing them there in a particular order but that does not change how they are shown in the web UI.  It seems they are always sorted alphabetically.

p.s. What are you using the timestamp for, I don't see it in the gallery repo yet?

**Reply by The-Lum** [reply]
Hello @matthew16550,

Currently the timestamp file name is: `plantuml-SNAPSHOT-timestamp.lock`.

But You can rename the file, as your convenience:
- `plantuml-SNAPSHOT.timestamp.lock`
- `plantuml-SNAPSHOT.timestamp`
- `plantuml-SNAPSHOT.timestamp.txt`

I will plan to use this file to control the download or not of the `plantuml-SNAPSHOT.jar`... _WIP...._

Regards.

**Reply by matthew16550** [reply]
Thanks @The-Lum,

I wonder if it might be cleaner to get the timestamp from the REST api?

This is the timestamp of the commit that is linked to the snapshot:

`curl -s https://api.github.com/repos/plantuml/plantuml/releases/tags/snapshot | jq -r .created_at`

And `jq -r .published_at` is when the snapshot was published.

**Comment by The-Lum** [comment]
FYI @matthew16550, @arnaudroques,

The last Snapshot is not the last Snapshot...
When several PR are merged, it seems that occurs this issue.

See [last Snapshot](https://github.com/plantuml/plantuml/releases/tag/snapshot): it is from 7f29798  (#776 / action [328](https://github.com/plantuml/plantuml/actions/runs/1465766982)) and not from 2098d24 of the last action (#775 / action [329](https://github.com/plantuml/plantuml/actions/runs/1465768455)).

**Reply by The-Lum** [reply]
Ah! I see, is that action [328](https://github.com/plantuml/plantuml/actions) lasted longer than action [329](https://github.com/plantuml/plantuml/actions)...
Then the Snapshot of 328 has overwritten the Snapshot of 329.
QED.

- How to counter that?

**Reply by arnaudroques** [reply]
Well, I think I have to wait for an action to finish before merging another PR :-)
That's the easiest solution.

**Reply by matthew16550** [reply]
I looked into this and have not found a simple automated solution.  Manually waiting before merging the next PR does seem the best approach for now 😢.

GH actions does not currently support sequential running of workflows in branch commit order.  The [concurrency](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#concurrency) option would probably solve this except it cancels other pending runs (https://github.com/github/feedback/discussions/5435).  For CI on the main branch we really should build each commit so do not want pending runs cancelled.

Tried some ideas around a job after the build that has `concurrency` set and searches runs in commit order for which jars should become the snapshot.  Or a separate workflow that decides which commit to build next.  But they are all complicated and probably have strange edge cases so we would be in a messier situation than now.

Hopefully GH will add something that solves it for us.

**Comment by The-Lum** [comment]
Hello @matthew16550 and all,

Here is an open question about CI:
- What is the reason to make the `CI` on `create`?
https://github.com/plantuml/plantuml/blob/29e652a3b074236019232a2b042554846f494ce2/.github/workflows/ci.yml#L4
- It is necessary or not?
_Especially on fork repo., each type I create a new branch the CI runs on my local repo.... and that make time..._
_(is it to be sure that it compiles before a first modification..., But in any case, the CI occurs also when you create a #PR)_

Regards.

**Reply by matthew16550** [reply]
The current release process is triggered by creating a tag.  It was not obvious how to trigger specifically on tags because `create:` seems to not accept any arguments.

Those extra builds had been bugging me too and you prompted more googling. Using `on.push.tags` might solve it.

But I think #769 will remove the need to trigger on tags, I will start looking at that today.

EDIT: Relevant forum topic https://github.community/t/how-to-run-github-actions-workflow-only-for-new-tags/16075

**Reply by The-Lum** [reply]
Thank you very much @matthew16550, it's perfectly clear...
_I also discover GitHub and the associated doc. of GitHub actions..._
